crates/red_knot/tests/file_watching.rs

```diff
@@ -0,0 +1,590 @@
+#![allow(clippy::disallowed_names)]
+
+use std::time::Duration;
+
+use anyhow::{anyhow, Context};
+
+use red_knot::db::RootDatabase;
+use red_knot::watch;
+use red_knot::watch::{directory_watcher, Watcher};
+use red_knot::workspace::WorkspaceMetadata;
+use red_knot_module_resolver::{resolve_module, ModuleName};
+use ruff_db::files::system_path_to_file;
+use ruff_db::program::{ProgramSettings, SearchPathSettings, TargetVersion};
+use ruff_db::source::source_text;
+use ruff_db::system::{OsSystem, SystemPath, SystemPathBuf};
+use ruff_db::Upcast;
+
+struct TestCase {
+    db: RootDatabase,
+    watcher: Option<Watcher>,
+    changes_receiver: crossbeam::channel::Receiver<Vec<watch::ChangeEvent>>,
+    temp_dir: tempfile::TempDir,
+}
+
+impl TestCase {
+    fn workspace_path(&self, relative: impl AsRef<SystemPath>) -> SystemPathBuf {
+        SystemPath::absolute(relative, self.db.workspace().root(&self.db))
+    }
+
+    fn root_path(&self) -> &SystemPath {
+        SystemPath::from_std_path(self.temp_dir.path()).unwrap()
+    }
+
+    fn db(&self) -> &RootDatabase {
+        &self.db
+    }
+
+    fn db_mut(&mut self) -> &mut RootDatabase {
+        &mut self.db
+    }
+
+    fn stop_watch(&mut self) -> Vec<watch::ChangeEvent> {
+        if let Some(watcher) = self.watcher.take() {
+            // Give the watcher some time to catch up.
+            std::thread::sleep(Duration::from_millis(10));
```

Comment:
Is this unavoidable? What kind "catching up" is needed that can't be expressed explicitly via a method that waits for the appropriate conditions? Sleeping in tests for an arbitrary amount of time is never good; inevitably it ends up being too short on some system and leading to flakiness, and wastes test-running time on every other system.

Reply:
> Is this unavoidable?

I couldn't figure out an alternative. I hate it too. 

There are a couple of things involved:

* We have to wait for the platform-specific syscall (`inotify`, `fsevents`, ...) to inform us about the file change
* The syscall doesn't happen on the main thread. So we have to wait for the file watcher, then the debounce threads to wake up and handle the event

I considered using a spin-loop that waits until the first event comes in, but even that's not guaranteed and risks dead-locking. So we would need something more complicated that avoids dead-locks as well (another debounce implementation :laughing:)

I'm open to any suggestions.

Reply:
I don't think we need any condvar for this. We can try to read from the `changes_receiver` (using the `select!` macro) and use a high timeout if no message is received in that time. But this is still not perfect. What if we receive a first batch of changes but then miss the last change message because it was a bit slower? 

I don't think a TODO is warranted because I don't plan to do anything about it unless the tests prove to be flaky. I also rather prefer the 10ms slow down than any complex system that only is prone to other kind of errors and is further away to what we use in production.  I'm not worried about the test performance. Ten ms isn't long enough to significantly slow down tests.